### PR TITLE
Fixed Configurable HadoopOutputFormat (FLINK-1828)

### PR DESCRIPTION
**Pull request by fpompermaier**
See https://issues.apache.org/jira/browse/FLINK-1828

### CODE REVIEW

File: flink-java/src/main/java/org/apache/flink/api/java/hadoop/mapreduce/HadoopOutputFormatBase.java [hunk 1]
```diff
@@ -66,6 +67,8 @@ public HadoopOutputFormatBase(org.apache.hadoop.mapreduce.OutputFormat<K, V> map
 	@Override
 	public void configure(Configuration parameters) {
 		// nothing to do
```
Comment (fhueske) [inline flink-java/src/main/java/org/apache/flink/api/java/hadoop/mapreduce/HadoopOutputFormatBase.java]: The comment should be removed.

File: flink-staging/flink-hbase/src/test/java/org/apache/flink/addons/hbase/example/HBaseWriteExample.java [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+		
+		DataSet<Tuple2<String, Integer>> counts = 
+				// split up the lines in pairs (2-tuples) containing: (word,1)
+				text.flatMap(new Tokenizer())
+				// group by the tuple field "0" and sum up tuple field "1"
+				.groupBy(0)
+				.sum(1);
+
+		// emit result
+//		if(fileOutput) {
```
Comment (fhueske) [inline flink-staging/flink-hbase/src/test/java/org/apache/flink/addons/hbase/example/HBaseWriteExample.java]: the `if` statement should be completely removed.

File: flink-staging/flink-hbase/src/test/java/org/apache/flink/addons/hbase/example/HBaseWriteExample.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+				@Override
+				public Tuple2<Text, Mutation> map(Tuple2<String, Integer> t) throws Exception {
+					reuse.f0 = new Text(t.f0);
+					Put put = new Put(t.f0.getBytes());
+					put.add(CF_SOME, Q_SOME, Bytes.toBytes(t.f1));
+					reuse.f1 = put;
+					return reuse;
+				}
+			}).output(new HadoopOutputFormat<Text, Mutation>(new TableOutputFormat<Text>(), job));
+//		} else {
```
Comment (fhueske) [inline flink-staging/flink-hbase/src/test/java/org/apache/flink/addons/hbase/example/HBaseWriteExample.java]: `else` branch not necessary

**Comment by fhueske** [comment]
Looks good to me, except for some comments that should be removed.

Regarding the `mapred.output.dir` parameter I am not sure whether this is generally expected for all Hadoop OutputFormats or only required for file-based OutputFormats. 
I would keep it for now and open a JIRA to investigate the issue and fix it if necessary.

**Comment by fpompermaier** [comment]
Removed comments and commented code as suggested by Fabian. Do I have also to create a JIRA ticket about mapred.output.dir parameter? I think that it can be defaulted to the Flink temp directory or flinkTempDir/hadoop/job-id

**Comment by mxm** [comment]
Is the Hadoop configuration specified in the flink-conf.yaml loaded? If we set `mapred.output.dir` then we should check for an existing config entry beforehand. Otherwise, we overwrite Hadoop configuration values. Like @fhueske suggested, please open a JIRA for investigation.

**Comment by fpompermaier** [comment]
Ok, I created this issue (https://issues.apache.org/jira/browse/FLINK-1834) about the mapred.output.dir

File: flink-staging/flink-hbase/src/test/resources/hbase-site.xml [hunk 4]
```diff
@@ -22,14 +22,13 @@
 -->
 <configuration>
 
+    <!--
```
Comment (fhueske) [inline flink-staging/flink-hbase/src/test/resources/hbase-site.xml]: Are these mandatory parameters to use HBase?
Otherwise, we should remove them.

File: flink-staging/flink-hbase/pom.xml [hunk 5]
```diff
@@ -112,6 +112,12 @@ under the License.
 				</exclusion>
 			</exclusions>
 		</dependency>
+		<dependency>
```
Comment (fhueske) [inline flink-staging/flink-hbase/pom.xml]: Why did you add this dependency? 
There are no additional tests that would require it, right?

**Comment by fhueske** [comment]
Looks good, except two minor things. 
Once these are resolved I would merge it and also backport it to the 0.8 branch.

File: flink-staging/flink-hbase/pom.xml (hunk 5 above)
Comment (fpompermaier) [inline flink-staging/flink-hbase/pom.xml]: It is needed if you want to use the HBase TableOutputFormat

File: flink-staging/flink-hbase/src/test/resources/hbase-site.xml (hunk 4 above)
Comment (fpompermaier) [inline flink-staging/flink-hbase/src/test/resources/hbase-site.xml]: I think you can remove the hbase-site.xml file. It is required only if you have hbase settings different from the default ones. Also the log4j.properties could be removed

File: flink-staging/flink-hbase/pom.xml (hunk 5 above)
Comment (fhueske) [inline flink-staging/flink-hbase/pom.xml]: But why is it in test scope then?

File: flink-staging/flink-hbase/src/test/resources/hbase-site.xml (hunk 4 above)
Comment (fhueske) [inline flink-staging/flink-hbase/src/test/resources/hbase-site.xml]: OK, let's remove them.

File: flink-staging/flink-hbase/pom.xml (hunk 5 above)
Comment (fpompermaier) [inline flink-staging/flink-hbase/pom.xml]: Because it is not required unless you want to write into HBase,  for the read it's not required

Comment (fhueske) [inline flink-staging/flink-hbase/pom.xml]: But putting it into test scope is not a proper solution to solve the issue. 
IMO, it should be either put in the regular scope such that it can be used at runtime or we put it into a comment and add a line explaining why we did that.

Comment (StephanEwen) [inline flink-staging/flink-hbase/pom.xml]: Is the HBase server dependency really required for any client that wants to write into HBase? This seems like a pretty bad design on the HBase side.

Can you tell us what fails when you omit this dependency?

Comment (fpompermaier) [inline flink-staging/flink-hbase/pom.xml]: Unfortunately the TableInputFormat and TableOutputFormat are in the server jar.
For the read we've reimplemented it to make it more robust so we don't need that jar, but for the output it is indeed required.

Comment (StephanEwen) [inline flink-staging/flink-hbase/pom.xml]: Fair enough. Then the dependency should not be in test scope, but in the default scope, so users get this dependency into their fat jar as well when using the HBase output format. May be worth to define a few exclusions, though, to not get the complete tail of transitive HBase dependencies (I think that even includes JRuby and so on)

Comment (fpompermaier) [inline flink-staging/flink-hbase/pom.xml]: Could you do that?

Comment (fhueske) [inline flink-staging/flink-hbase/pom.xml]: I do not have a HBase setup here. 
Could you try to exclude all dependencies of hbase-server and add them until it works? I hope the TableInputFormat and TableOutputFormat have not too many external dependencies.

Comment (fpompermaier) [inline flink-staging/flink-hbase/pom.xml]: Ok, I hope to be able to do it before this evening!

**Comment by fhueske** [comment]
This PR was split into PR #632 and PR #633
